### ISSUE TITLE
Measure shared memory usage of Artery

### ISSUE THREAD

**Issue by 2m** [akka/akka#21923]
So we can arrive at a useful recommendation for the shared memory mount size.

**Comment by guidomedina** [comment]
I had a problem few days ago with an UAT environment that wouldn't allow 10 JVMs to start up on 4 VMs each with 2g of RAM, so I had to create a tunnable shared driver for different environment, the formula I have observed is the following:

Each node will reserve `aeron.term.buffer.length` * 12, that said, its default value is 16m hence occupying 192m per JVM in the cluster per shared driver, thats is to say that if you have 10 JVMs spread out in 4 physical machines and each machine with a shared driver, each machine will 192m * 10 JVMs which is 1920m, making each machine to require at least 4g+ of RAM.

Why 4g+? is not only 1920m, it is 1920m + some overhead > 2g, that IMHO was too much, hence I tuned down  `aeron.term.buffer.length` to 4m hence making it possible to run 10 JVMs among 4 physical machines (in my case virtual machines) with 2g RAM.

Here is my tunnable Media Driver Java main:

```
import com.typesafe.config.Config;
import io.aeron.driver.MediaDriver;
import org.agrona.concurrent.BackoffIdleStrategy;
import org.agrona.concurrent.IdleStrategy;
import org.agrona.concurrent.SigIntBarrier;

import java.io.File;

import static com.typesafe.config.ConfigFactory.parseFile;
import static java.util.concurrent.TimeUnit.MICROSECONDS;
import static org.jctools.util.Pow2.roundToPowerOfTwo;

public class MediaDriverMain {

  public static IdleStrategy createIdleStrategy(int idleCpuLevel) {
    return new BackoffIdleStrategy(1100 * idleCpuLevel - 1000, 5 * idleCpuLevel, 1, MICROSECONDS.toNanos(280 - 30 * idleCpuLevel));
  }

  public static void main(String[] args) {
    final Config config = parseFile(new File(args[0])).getConfig("aeron");
    final int idleCpuLevel = config.getInt("idle_cpu_level");
    if (idleCpuLevel < 1 || idleCpuLevel > 9) {
      System.out.println("Invalid IDLE_CPU_LEVEL, valid values are from 1 to 9.");
      return;
    }
    final int termBufferSize = roundToPowerOfTwo(config.getInt("term_buffer_size"));
    if (termBufferSize < 2 || termBufferSize > 16) {
      System.out.println("Invalid TERM_BUFFER_SIZE, valid values are 2, 4, 8 or 16.");
      return;
    }
    System.out.println("Starting media driver with IDLE_CPU_LEVEL = " + idleCpuLevel + " and TERM_BUFFER_SIZE = " + termBufferSize);
    try (MediaDriver ignored = MediaDriver.launch(new MediaDriver.Context().
      aeronDirectoryName("/dev/shm/aeron").
      receiverIdleStrategy(createIdleStrategy(idleCpuLevel)).
      senderIdleStrategy(createIdleStrategy(idleCpuLevel)).
      publicationTermBufferLength(termBufferSize * 1024 * 1024).
      dirsDeleteOnStart(true))) {
      System.out.println("Media driver started successfully - waiting for a SIGTERM to exit.");
      new SigIntBarrier().await();
    } finally {
      System.out.println("Media driver has been shutdown.");
    }
  }
}
```

It expects a config like the following in the passed config file:

```
aeron {
  idle_cpu_level = 2
  term_buffer_size = 4
}
```

There isn't really a hard requirement for `aeron.term.buffer.length`, its wiki says it is only meant to keep the publisher ahead of the consumer, so, higher CPU usage could mean less RAM required, so far in my case for a low latency scenario is working fine for several hundreds of messages/sec with that shown config.

**Comment by guidomedina** [comment]
To add some background, by default most Linux dist will mount half of the RAM as shared memory, hence my explanation implicating amount of RAM and Media Driver tuning, in most servers you don't have to do anything, in fact adjusting shared memory wouldn't be recommended but instead adjust the Media Driver.

**Comment by guidomedina** [comment]
`aeron.term.buffer.length` parameter needs to be part of `akka-remote` config when running the embedded driver, I suggest to tune it down to `8m` or even `4m`

**Comment by patriknw** [comment]
Thanks for valuable input, @guidomedina 
It should be possible to define `aeron.term.buffer.length` and all other Aeron properties as system properties (`-D jvm flags) and it will be picked up by the embedded media driver. We prefer to avoid duplicating all properties.

**Comment by guidomedina** [comment]
I agree with you but notice this one is an exception, once shared memory is filled a node cannot join the cluster which might be tricky for people testing clusters.

By adding Aeron you guys raised the level of expertise a developer needs to work with an akka cluster, I don't expect a production environment using the embedded driver, that said, it might be reasonable to tune down this special parameter.

But it can be a double edge sword to do it because it will hide such fact from a developer, making the developer think that it is ok to start embedded drivers all over.

**Comment by patriknw** [comment]
It might turn out that config of term buffer is important enough. We'll see.
